docs/proposals/20220603-kubernetes-vendor.md

```diff
@@ -0,0 +1,235 @@
+$ envd build
+$ envd push
+$ envd run --env test --image test
+
+or 
+
+$ envd context create --name test --builder-name test --use --builder kube-pod --runner kubernetes --runner-config kube-pod-config.yaml
+$ envd up
+```
+
+### Implementation details/Notes/Constraints
+
+![](../images/proposals/k8s.png)
+
+#### envd contexts
+
+envd uses buildkitd container in the local docker host to build the images by default. `context` command will be introduced to support remote build.
+
+```
+NAME:
+   envd context - Manage envd contexts
+
+USAGE:
+   envd context command [command options] [arguments...]
+
+COMMANDS:
+   create   Create envd context
+   ls       List envd contexts
+   rm       Remove envd context
+   use      Use the specified envd context
+   help, h  Shows a list of commands or help for one command
+
+OPTIONS:
+   --help, -h  show help (default: false)
+```
+
+The structure of the envd context is:
+
+```go
+type Context struct {
+	Name          string      `json:"name,omitempty"`
+	Builder       BuilderType `json:"builder,omitempty"`
+	BuilderSocket string      `json:"builder_socket,omitempty"`
+
+	Runner     RunnerType `json:"runner,omitempty"`
+	DockerHost *string    `json:"docker_host,omitempty"`
+	KubeConfig *string    `json:"kube_config,omitempty"`
+}
+
+type BuilderType string
+
+const (
+	BuilderTypeDocker     BuilderType = "docker-container"
+	BuilderTypeKubernetes BuilderType = "kube-pod"
+)
+
+type RunnerType string
+
+const (
+	RunnerTypeDocker     RunnerType = "docker"
+	RunnerTypeKubernetes RunnerType = "kubernetes"
+)
+```
+
+The default envd context is:
+
+```go
+Context{
+    Name:          "default",
+    Builder:       BuilderTypeDocker,
+    BuilderSocket: "unix:///var/run/docker.sock",
+    Runner:        RunnerTypeDocker,
+    DockerHost:    nil,
+    KubeConfig:    nil,
+}
+```
+
+Users may use `envd context create` to create new context and use it:
+
+```
+$ envd context create --name my-context --builder-type docker-container --runner-type docker --docker-host unix:///var/run/docker.sock
+```
+
+#### Entrypoint
+
+Currently, entrypoint is set during the container creation:
+
+```go
+	config := &container.Config{
+		Entrypoint: []string{
+			"tini",
+			"--",
+			"bash",
+			"-c",
+		},
+	}
+	config.Entrypoint = append(config.Entrypoint,
+		entrypointSH(g, config.WorkingDir, sshPort))
+```
+
+To support the Kubernetes, envd should generate entrypoint and cmd for the image during the build process, instead of in the runtime.
+
+#### Bi-directional filesystem sync
+
+A per-pod sidecar is introduced to sync files between host and Kubernetes pod. [syncthing](https://docs.syncthing.net/intro/getting-started.html) will run in the per-pod sidecar.
+
+The sidecar design does not involve changing the image.
+
+#### Port forwarding
+
+There are several ports will be used in envd:
+
+- sshd server port. envd randomly selects a host port for sshd.
```

Comment:
Just for curiosity, do we need ssh into the host or the container?

Reply:
If port-forwarding is usable under this scenario, I think we could forward a local port to the sshd port inside the container, then we could ssh into the container directly. 

In addition, even if we selected a host port, the host might not be public accessible, hence we could not connect to the host.